doc/developer/design/20220728_persisting_introspection_data.md

```diff
@@ -0,0 +1,235 @@
+
+- Crashing replica
+
+If a replica crashes (for example due to being out-of-memory), the introspection data can’t reliably be saved by the user before the crash happens, as it happened unexpectedly. It is important for the user to see the introspection data for a postmortem analysis.
+
+- Restarting replica
+
+Related to the crashed replicas, we have to consider what happens when a replica restarts, since Kubernetes (potentially with some back-off) does this automatically. If we want to analyze the fault scenario it is that the user can distinguish from which lifetime the data stems. On the other hand, keeping around a lot of data from a long defunct replica might distract the user.
+
+- In future versions: Using introspection for autoscaling
+
+At some point in the distant future, we might use this data to automatically detect an overloaded replica and migrate the instance to a bigger instance automatically.
+
+## Current state
+
+### M1: Replica directed queries
+
+In M1, we implemented [replica targeted queries](https://github.com/MaterializeInc/materialize/pull/12568) queries. They use the special variable `cluster` and `cluster_replica` to direct a future `SELECT` to a specific replica. Using this technique, it is possible to query the introspection information of a given replica.
+
+```rust
+CREATE CLUSTER test
+  REPLICA replica_a (SIZE '1'),
+  REPLICA replica_b (SIZE '2');
+SET cluster = test;
+SET cluster_replica = replica_a;
+SELECT * from mz_materializations;rmissions that must be reviewed by an App Manager before you can inst
+```
+
+Setting this variable will have an effect on all queries, not only those that target introspection sources.
+
+
+## Difficulties in offering an aggregated source
+
+From a pure SQL perspective, the obvious schema to present the cross-replica introspection information is to extend the introspection tables with a replica id column.
+
+For example `mz_scheduling_histograms` would get an additional column `replica_id`.
+
+```
+worker    slept_for    requested    count
+-----------------------------------------
+...
+```
+
+Would become
+
+```
+replica_id     worker    slept_for    requested    count
+---------------------------------------------------------
+...
+```
+
+However, if done naively the way queries work in materialize requires a fixed timestamp at which time the results of the queries are reported. But if one replica is lagging or otherwise unable to provide logging data, the query would either report out of date data (when settling for the least timestamp where all replicas have provided data) or block for a long time (when choosing a timestamp closer to the actual time). See the design question “Expose an aggregate sources or source per replica” as well as the [discussion in the Github](https://github.com/MaterializeInc/materialize/pull/13340#issuecomment-1180628374) PR for more technical details.
+
+## The zombie replica problem
```

Comment:
Is this really a problem? What happens if another replica tries to write into the same collection? Will persist notice this?

Reply:
I don't exactly know. I think what happens is that the `compare_and_append` will fail because someone else incremented the timestamp. Then computed could actually react and crash.